src/transformers/feature_extraction_utils.py

```diff
@@ -175,25 +176,55 @@ def as_tensor(value):
         return self
 
     @torch_required
-    # Copied from transformers.tokenization_utils_base.BatchEncoding.to with BatchEncoding->BatchFeature
-    def to(self, device: Union[str, "torch.device"]) -> "BatchFeature":
+    def to(self, *args, **kwargs) -> "BatchFeature":
         """
-        Send all values to device by calling `v.to(device)` (PyTorch only).
+        Send all values to device by calling `v.to(*args, **kwargs)` (PyTorch only). This should support casting in
+        different `dtypes` and sending the `BatchFeature` to a different `device`.
 
         Args:
-            device (`str` or `torch.device`): The device to put the tensors on.
+            args (`Tuple`):
+                Will be passed to the `to(...)` function of the tensors.
+            kwargs (`Dict`, *optional*):
+                Will be passed to the `to(...)` function of the tensors.
 
         Returns:
             [`BatchFeature`]: The same instance after modification.
         """
-
-        # This check catches things like APEX blindly calling "to" on all inputs to a module
-        # Otherwise it passes the casts down and casts the LongTensor containing the token idxs
-        # into a HalfTensor
-        if isinstance(device, str) or is_torch_device(device) or isinstance(device, int):
-            self.data = {k: v.to(device=device) for k, v in self.data.items()}
-        else:
-            logger.warning(f"Attempting to cast a BatchFeature to type {str(device)}. This is not supported.")
+        import torch  # noqa
+
+        new_data = {}
+        # We cast only floating point tensors to avoid issues with tokenizers casting `LongTensor` to `FloatTensor`
+        for k, v in self.items():
+            # check if v is a floating point
+            if torch.is_floating_point(v):
+                # cast and send to device
+                new_data[k] = v.to(*args, **kwargs)
+            else:
+                # just send to device
+                device = kwargs.pop("device", None)
+                # Check if the args are a device or a dtype
+                if device is None and len(args) > 0:
+                    # device should be always the first argument
+                    arg = args[0]
+                    if is_torch_dtype(arg):
+                        # Ignore the dtype
+                        logger.warning(
+                            "Attempting to cast a non-floating point element of BatchFeature to `dtype`"
+                            f" {str(arg)}. This is not supported."
+                        )
+                    elif isinstance(arg, str) or is_torch_device(arg) or isinstance(arg, int):
+                        device = arg
+                    else:
+                        # it's something else
+                        raise ValueError(
+                            f"Attempting to cast a BatchFeature to type {str(arg)}. This is not supported."
+                        )
```

Comment:
The analysis of the args/kwargs is done at every step of the loop. It's not very efficient when it could be done once and for all at the beginning. Also, no warning should be sent for int tensors IMO, especially as it's going to be triggered a lot of time!

Reply:
Thanks! Indeed the several loops are not needed